library/Booster/Pattern/Rewrite.hs

```diff
@@ -372,8 +372,12 @@ applyRule pat@Pattern{ceilConditions} rule = runRewriteRuleAppT $ do
         RewriteRuleAppT (RewriteT io (RewriteFailed k)) (Maybe a)
     checkConstraint onUnclear onBottom p = do
         RewriteConfig{definition, llvmApi, smtSolver, doTracing} <- lift $ RewriteT ask
-        (simplified, _traces, _cache) <-
-            simplifyConstraint (castDoTracingFlag doTracing) definition llvmApi smtSolver mempty p
+        oldCache <- lift . RewriteT . lift $ get
+        (simplified, _traces, cache) <-
+            simplifyConstraint (castDoTracingFlag doTracing) definition llvmApi smtSolver oldCache p
+        -- update cache
+        lift . RewriteT . lift . modify $ const cache
+        -- TODO should we keep the traces? Or only on success?
```

Comment:
I think it makes sense to keep both success and failure traces.

Reply:
I suggest we don't bother with the traces in this PR, as I have a similar change pending in https://github.com/runtimeverification/hs-backend-booster/pull/504. Perhaps it would be best to batch these together so that we have to run the KEVM/Kontrol performance regression suites less times.